### PR TITLE
Improve reflection_pad2d lowering for dynamic shapes

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110988

Fixes https://github.com/pytorch/pytorch/issues/110696

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by ezyang** [comment]
Still needs tests, but a review will still be helpful

File: torch/_inductor/index_propagation.py [hunk 1]
```diff
@@ -54,7 +54,9 @@ def identity(value: Any) -> Any:
 
     @staticmethod
     def constant(value: Union[int, float, bool], dtype: torch.dtype) -> TypedExpr:
-        if is_boolean_dtype(dtype):
+        if isinstance(value, sympy.Expr):
+            expr = value
```
Comment (peterbell10) [inline torch/_inductor/index_propagation.py:58]: `ops.constant` shouldn't be called with `sympy.Expr`. Instead it should be `ops.index_expr`.

Comment (ezyang) [reply torch/_inductor/index_propagation.py:58]: So it turns out that it's easy enough to change the ops.constant call in the lowering to ops.index_expr, but now I am wondering, is there really any material difference between these two calls? Like, it's not like you're going to get worse code if you call index_expr with a constant, no?

Comment (peterbell10) [reply torch/_inductor/index_propagation.py]: Ideally I think they should be similar but in practice they are materially different. `ops.constant` will respect the dtype argument and generate code like `tl.full([1, 1], 256, tl.int32)` whereas `ops.index_expr` goes through the sympy expression printer which gives an integer literal `256` which ignores the specified dtype.

Comment (ezyang) [reply torch/_inductor/index_propagation.py]: in any case I fixed it, plz review lol

**Review by jansel — APPROVED** [review]

**Review by lezcano — APPROVED** [review]

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -3180,7 +3178,7 @@ def fn(idx):
         device=x.get_device(),
         dtype=x.get_dtype(),
         inner_fn=fn,
-        ranges=[*batch, sympy.Integer(h + top + bot), sympy.Integer(w + left + right)],
+        ranges=[*batch, sympy.sympify(h + top + bot), sympy.sympify(w + left + right)],
```
Comment (eellison) [inline torch/_inductor/lowering.py:3181]: why do we need `sympy.sympify` here ?

Comment (ezyang) [reply torch/_inductor/lowering.py:3181]: if h, top and bot happen to be all int, we won't return a sympy.Integer, but I noticed that in inductor it is sometimes load bearing to return a sympy expression, not a plain int

Comment (eellison) [reply torch/_inductor/lowering.py:3181]: It should be fine to return an int here

**Pull request closed without merging**